flight/mcu/lib/teensyShared/HistoryBufferFSM50.cpp

```diff
@@ -0,0 +1,323 @@
+            }
+            // potentially add back state to put us back into coast
+
+            if (TIME_I2MS(apogee_timer_) <
+                drogue_deploy_time_since_apogee_threshold) {
+                rocket_state_ = FSM_State::STATE_APOGEE;
+            } else {
+                rocket_state_ = FSM_State::STATE_DROGUE;
+                drogue_time_ = chVTGetSystemTime();
+            }
+            break;
+
+        case FSM_State::STATE_DROGUE_DETECT:
+            if (fabs(
+                    (*altitude_history_ptr_)
+                        .getCurrentSecondDerivativeAverage() -
+                    (*altitude_history_ptr_).getPastSecondDerivativeAverage()) >
+                drogue_acceleration_change_threshold_altimeter) {
+                rocket_state_ = FSM_State::STATE_DROGUE;
+                drogue_time_ = chVTGetSystemTime();
+                break;
+            }
+
+            else {
+                rocket_state_ = FSM_State::STATE_APOGEE;
+            }
+            break;
+
+        case FSM_State::STATE_DROGUE:
+            drogue_timer_ = chVTGetSystemTime() - drogue_time_;
+            if (TIME_I2MS(drogue_timer_) > refresh_timer) {
+                if (fabs((*IMU_acceleration_history_ptr_).getCurrentAverage() -
+                         (*IMU_acceleration_history_ptr_).getPastAverage()) >
+                    main_acceleration_change_threshold_imu) {
+                    rocket_state_ = FSM_State::STATE_MAIN_DETECT;
+                    break;
+                }
+            }
+
+            if (TIME_I2MS(drogue_timer_) <
+                main_deploy_time_since_drogue_threshold) {
+                rocket_state_ = FSM_State::STATE_DROGUE;
+
+            } else {
+                rocket_state_ = FSM_State::STATE_MAIN;
+                main_time_ = chVTGetSystemTime();
+            }
+            break;
+
+        case FSM_State::STATE_MAIN_DETECT:
+            if (fabs(
+                    (*altitude_history_ptr_)
+                        .getCurrentSecondDerivativeAverage() -
+                    (*altitude_history_ptr_).getPastSecondDerivativeAverage()) >
+                main_acceleration_change_threshold_altimeter) {
+                rocket_state_ = FSM_State::STATE_MAIN;
+                main_time_ = chVTGetSystemTime();
+                break;
+            } else {
+                rocket_state_ = FSM_State::STATE_DROGUE;
+            }
+            break;
+
+        case FSM_State::STATE_MAIN:
+            main_timer_ = chVTGetSystemTime() - main_time_;
+
+            if (fabs((*altitude_history_ptr_).getCurrentAverage() -
+                     (*altitude_history_ptr_).getPastAverage()) <
+                landing_altimeter_threshold) {
+                rocket_state_ = FSM_State::STATE_LANDED_DETECT;
+                landing_time_ = chVTGetSystemTime();
+                break;
+            }
+
+            if (TIME_I2MS(main_timer_) <
+                main_deploy_time_since_drogue_threshold) {
```

Comment:
This is the wrong threshold. main_deploy_time_since_drogue_threshold is how long until we force the state to go from drogue to main. We instead need to make a new threshold that is how long until we force the state from main to landed. This error appears in all 4 FSMs.